### DISCUSSION TITLE
No sound — any debugging advice?

### DISCUSSION THREAD

**Discussion by bberstel** [cardonabits/haxo-hw#56] [Q&A]
Hi,

I have unpacked my Haxophone today, and first of all: **thanks a lot** Javier for bringing this brilliant idea to life, and to the rest of us!

I am using a Raspberry Pi Zero W, and the pre-built image. I'm not much of a hardware guy, but I think I have done everything right. (Actually the hardest part was to find a computer with a SD port…) When powered, the Pi boots with no error message (except the one related to swap mentioned in another discussion, and which I understand can be safely ignored).

Yet, when I blow in the Haxophone, I get no sound in my headphones. I have tried the headphones on a CD player: they work.

I must say that I'm a bit clueless. As I wrote, I'm not much of a hardware guy, but I have experience in software. So if there is anything I can try to debug my situation, I'll do it…

Thanks in advance for any help.
Bruno

**Comment by jcard0na** [comment]
Hi Bruno,

Thanks for the words of support... even if you are getting no sound! Let's try to fix that! 😅
 
Based on your comment regarding error messages, it seems like you have a console, maybe by plugging a display and keyboard to the Pi?  If so, can you log in (user: `pi` pwd:`haxophone`) and print the haxo service logs? 
```
journalctl -u haxo 
```
That should tell us if the service started successfully.

Other useful logs:
```
aplay -l
```

Cheers,

**Reply by bberstel** [reply]
Thanks for your quick answer! I have to mention: although I'm eager to have my haxophone working, there is no dramatic urgency, so please remember that it's quite late in the evening and that we're in the holiday season. :)

I have run `journalctl` and indeed the service did not start successfully:
```
systemd[1]: Started haxophone.
create_midi_gadget.sh[…]: mkdir: cannot create directory XYZ (5 of them): File exists
create_midi_gadget.sh[…]: ln: failed to create symbolic link 'configs/c.1/midi.usb0': File exists
/usr/local/bin/create_midi_gadget.sh: 17: echo: echo: I/O error
systemd[1]: haxo.service: Control process exited, code=exited, status=1/FAILURE
systemd[1]: haxo.service: Failed with result 'exit-code'
systemd[1]: haxo.service: Scheduled restart job, restart counter is at 154.
systemd[1]: Stopped haxophone.
```
I have a number of such sequences of lines.

As for the `aplay -l` command, it lists two cards:
- card 0: MAX98357A
- card 1: vc4hdmi
Both say `Subdevices: 1/1` and `Subdevice #0: subdevice #0`.

However, the error messages in the execution of the `create_midi_gadget.sh` script may not be harmful, as it's not a problem not to be able to create a directory that already exists. But I suspect that since the last statement of the script failed, the whole script is considered failing, hence the `haxo.service: Failed with result 'exit-code'` message. I started adding a `true` statement at the end of the script, but I'm hitting the limits of my Pi knowledge, because when I rebooted the Pi, of course the statement wasn't there any longer.

However since the Pi is retrying to launch the service, the log changed a bit:
```
systemd[1]: Started haxophone.
create_midi_gadget.sh[…]: mkdir: cannot create directory XYZ (5 of them): File exists
create_midi_gadget.sh[…]: ln: failed to create symbolic link 'configs/c.1/midi.usb0': File exists
/usr/local/bin/create_midi_gadget.sh: 17: echo: echo: I/O error
haxo001[…]: fluidsynth: warning: SDL2 not initialized, SDL2 audio driver won't be usable
haxo001[…]: [timestamp INFO  haxo001::alsa] Found alsa card MAX98357A
haxo001[…]: synth created
haxo001[…]: [timestamp INFO  haxo001::midi]
haxo001[…]:     Available output ports:
haxo001[…]: [timestamp INFO  haxo001::midi] 0: Midi Through:Midi Through Port-0 14:0
haxo001[…]: [timestamp INFO  haxo001::midi] 1: f_midi:f_midi 24:0
haxo001[…]: [timestamp INFO  haxo001::midi] Picked the last one
haxo001[…]: Starting haxophone...
haxo001[…]: thread 'main' panicked at 'Failed to initialise pressure sensor: Io(Os { code: 121, kind: Uncategorized, message: "Remote I/O error" })', src/main.rs:92:49
haxo001[…]: note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
systemd[1]: haxo.service: Main process exited, code=killed, status=11/SEGV
systemd[1]: haxo.service: Failed with result 'signal'
systemd[1]: haxo.service: Consumed 7.327s CPU time.
systemd[1]: haxo.service: Scheduled restart job, restart counter is at 154.
systemd[1]: Stopped haxophone.
```

Best, B

**Reply by jcard0na** [reply]
Hi Bruno,

That is very useful.  There is a [known issue](https://github.com/cardonabits/haxo-rs/issues/22) with *restarting* the service.  But we need to know why the first attempt failed.  You could have just scrolled to the beginning of the log, but your changes to the script also worked.  Nice!

This seems to be the root cause, and it points to a hardware issue:
```
...
haxo001[…]: thread 'main' panicked at 'Failed to initialise pressure sensor: Io(Os { code: 121, kind: Uncategorized, message: "Remote I/O error" })', src/main.rs:92:49
...
```
I know you said you are not much of a hardware person, so I won't ask you to measure voltages, etc.  But can you visually inspect the [pressure sensor on the haxophone](https://github.com/cardonabits/haxo-hw/blob/main/docs/assembly.md#6-solder-pressure-sensor) for signs of damage?  There were two variants of the sensor installed, one has two ports and the other has only one.  The one in the pictures has two, do not pay attention to that.  But maybe it was unsoldered or missing altogether.

Alternatively, it could be an issue with the I2C bus on your Raspberry Pi.  Do you have a second one you could try?

Cheers,

**Reply by jcard0na** [reply]
BTW, to make the changes to the script permanent, you can disable overlay, change the script, and enable it again.
```
sudo raspi-config
Performance Options -> Overlay Filesystem -> Enable Overlay? [No|Yes]
```
That will fix the restart issue and eliminate the noise in the logs.

**Reply by bberstel** [reply]
Hi,

> There is a known issue with _restarting_ the service. But we need to know why the first attempt failed. You could have just scrolled to the beginning of the log

Yes I realised that only later. 😄 So this morning I had a look at the start of the log, just to be sure. The log contains three boots. The first one is dated a bit more than two hours before the current time, and successfully starts the `haxo` service, then stops it two hours later. The second boot happens in the subsequent minute, also successfully starts the service, and stops it after five minutes. The third boot seems to be "my" boot, and displays the pressure sensor initialisation error. Maybe that's all normal, just mentioning.

Regarding the pressure sensor: mine has only one port (so it is present). I can see no sign of damage on the solders or otherwise.

I unfortunately do not have another RPi. I googled how to test the I2C bus, I [found](https://learn.adafruit.com/scanning-i2c-addresses/raspberry-pi) that I had to install the `i2c-tools`, which requires Internet connection. But hey, the Pi is a W, so let's connect to wi-fi. (Which by the way could allow me to ssh to the Pi, must more convenient…) But I gave up as I couldn't work the errors out. I tried with `raspi-config`, which says `No wireless interface found`. I tried with `wpa_cli`, which says `Failed to connect to non-global ctrl_iframe: wlan0  error: No such file of directory`. Maybe the pre-built image does not enable it?

Here's where I am for now…

Best, B

**Reply by bberstel** [reply]
I have good news and bad news.

I just found another RPi at a shop nearby. I took exactly the same model (Zero W v1.1) and used it instead of the previous one. The good news is I no longer have the error in the log. That is, the log stops after `Starting haxophone...`. The bad news is that I still have no sound.

Other things I noticed: it takes much more time between the moment I type in the username & password, and the moment I get the shell prompt. More generally, the Pi seems less responsive. I tend to see a good sign there, as it is probably busy doing its haxophone job. Also, on my first try, I had strange i/o on the console (see photo). I rebooted to see whether I re-occurred but it did not (and still no sound).

![console](https://github.com/cardonabits/haxo-hw/assets/22963792/96c2eead-8f49-4b30-89fb-29c0b36d9c53)

Of course, I'm both happy and unhappy with this. But rather more happy as there is progress. I'd be glad to try anything else to debug my problem further.

Best, B

**Reply by jcard0na** [reply]
Oh, wow, lots of information here.  I do not have a solution, but I concur with you that there is progress... 😄 

Things that come to mind:

1. My image has WiFi disabled for better performance.  If you need to enable it, remove this line from `/boot/config.txt` (You will have to disable the overlay, and make the boot partition writable, both via `raspi-config`)
```
dtoverlay=disable-wifi
```
2. The "old" haxo logs.  I was not aware that there were two old boot logs in the pre-installed image!  I thought they would be wiped out by the overlay system, but apparently not.  I just booted my haxophone and saw the same old logs as you did, identified by PIDs 504 and 508.  TIL 🔦 
```
pi@haxophone:~$ journalctl -u haxo --no-pager                                                                                                                      
-- Journal begins at Thu 2023-12-21 16:22:32 GMT, ends at Thu 2023-12-21 20:36:13 GMT. --                                                                          
Dec 21 18:26:25 haxophone systemd[1]: Started haxophone.                                                                                                           
Dec 21 18:26:32 haxophone haxo001[508]: fluidsynth: warning: SDL2 not initialized, SDL2 audio driver won't be usable                                               
Dec 21 18:26:32 haxophone haxo001[508]: [2023-12-21T18:26:32Z INFO  haxo001::alsa] Found alsa card MAX98357A                                                       
Dec 21 18:26:42 haxophone haxo001[508]: Synth created                                                                                                              
Dec 21 18:26:43 haxophone haxo001[508]: [2023-12-21T18:26:43Z INFO  haxo001::midi]                                                                                 
Dec 21 18:26:43 haxophone haxo001[508]:     Available output ports:                                                                                                
Dec 21 18:26:43 haxophone haxo001[508]: [2023-12-21T18:26:43Z INFO  haxo001::midi] 0: Midi Through:Midi Through Port-0 14:0                                        
Dec 21 18:26:43 haxophone haxo001[508]: [2023-12-21T18:26:43Z INFO  haxo001::midi] 1: f_midi:f_midi 24:0                                                           
Dec 21 18:26:43 haxophone haxo001[508]: [2023-12-21T18:26:43Z INFO  haxo001::midi] Picked the last one                                                             
Dec 21 18:26:43 haxophone haxo001[508]: Starting haxophone...                                                                                                      
Dec 21 20:28:17 haxophone systemd[1]: Stopping haxophone...                                                                                                        
Dec 21 20:28:18 haxophone systemd[1]: haxo.service: Succeeded.                                                                                                     
Dec 21 20:28:18 haxophone systemd[1]: Stopped haxophone.                                                                                                           
Dec 21 20:28:18 haxophone systemd[1]: haxo.service: Consumed 20min 30.725s CPU time.                                                                               
-- Boot 23f5ccd22f674322a70604d7f4547235 --                                                                                                                        
Dec 21 20:29:15 haxophone systemd[1]: Started haxophone.                                                                                                           
Dec 21 20:29:21 haxophone haxo001[504]: fluidsynth: warning: SDL2 not initialized, SDL2 audio driver won't be usable                                               
Dec 21 20:29:21 haxophone haxo001[504]: [2023-12-21T20:29:21Z INFO  haxo001::alsa] Found alsa card MAX98357A                                                       
Dec 21 20:29:32 haxophone haxo001[504]: Synth created                                                                                                              
Dec 21 20:29:33 haxophone haxo001[504]: [2023-12-21T20:29:33Z INFO  haxo001::midi]                                                                                 
Dec 21 20:29:33 haxophone haxo001[504]:     Available output ports:                                                                                                
Dec 21 20:29:33 haxophone haxo001[504]: [2023-12-21T20:29:33Z INFO  haxo001::midi] 0: Midi Through:Midi Through Port-0 14:0                                        
Dec 21 20:29:33 haxophone haxo001[504]: [2023-12-21T20:29:33Z INFO  haxo001::midi] 1: f_midi:f_midi 24:0                                                           
Dec 21 20:29:33 haxophone haxo001[504]: [2023-12-21T20:29:33Z INFO  haxo001::midi] Picked the last one                                                             
Dec 21 20:29:33 haxophone haxo001[504]: Starting haxophone...                                                                                                      
Dec 21 20:34:41 haxophone systemd[1]: Stopping haxophone...                                                                                                        
Dec 21 20:34:42 haxophone systemd[1]: haxo.service: Succeeded.                                                                                                     
Dec 21 20:34:42 haxophone systemd[1]: Stopped haxophone.                                                                                                           
Dec 21 20:34:42 haxophone systemd[1]: haxo.service: Consumed 3min 8.759s CPU time.                                                                                 
-- Boot a812594f58a640a781fd7248b3c23ed5 --                                                                                                                        
Dec 21 20:35:30 haxophone systemd[1]: Started haxophone.                                                                                                           
Dec 21 20:35:37 haxophone haxo001[558]: fluidsynth: warning: SDL2 not initialized, SDL2 audio driver won't be usable                                               
Dec 21 20:35:37 haxophone haxo001[558]: [2023-12-21T20:35:36Z INFO  haxo001::alsa] Found alsa card MAX98357A                                                       
Dec 21 20:35:47 haxophone haxo001[558]: Synth created                                                                                                              
Dec 21 20:35:48 haxophone haxo001[558]: [2023-12-21T20:35:48Z INFO  haxo001::midi]                                                                                 
Dec 21 20:35:48 haxophone haxo001[558]:     Available output ports:                                                                                                
Dec 21 20:35:48 haxophone haxo001[558]: [2023-12-21T20:35:48Z INFO  haxo001::midi] 0: Midi Through:Midi Through Port-0 14:0                                        
Dec 21 20:35:48 haxophone haxo001[558]: [2023-12-21T20:35:48Z INFO  haxo001::midi] 1: f_midi:f_midi 24:0                                                           
Dec 21 20:35:48 haxophone haxo001[558]: [2023-12-21T20:35:48Z INFO  haxo001::midi] Picked the last one                                                             
Dec 21 20:35:48 haxophone haxo001[558]: Starting haxophone... 
```
3. The second Zero W.  Yes, this is good news.  If the haxo process is running on the second Zero, you can confirm that the pressure sensor was initialized correctly.  To check that the process is running you can run `systemctl status haxo` and see that the process has been active for at least a few minutes...
```
pi@haxophone:~$ systemctl status haxo --no-pager                                                                                                                   
● haxo.service - haxophone                                                                                                                                         
     Loaded: loaded (8;;file://haxophone/etc/systemd/system/haxo.service/etc/systemd/system/haxo.service8;;; enabled; vendor preset: enabled)                      
     Active: active (running) since Thu 2023-12-21 20:35:30 GMT; 10min ago                                                                                         
    Process: 446 ExecStartPre=/usr/local/bin/create_midi_gadget.sh (code=exited, status=0/SUCCESS)                                                                 
   Main PID: 558 (haxo001)                                                                                                                                         
      Tasks: 3 (limit: 395)                                                                                                                                        
        CPU: 7min 4.502s                                                                                                                                           
     CGroup: /system.slice/haxo.service                                                                                                                            
             └─558 /usr/local/bin/haxo001 --notemap-file /usr/share/haxo/notema…                                                                                   
                                                                                                                                                                   
Dec 21 20:35:30 haxophone systemd[1]: Started haxophone.                                                                                                           
Dec 21 20:35:37 haxophone haxo001[558]: fluidsynth: warning: SDL2 not initi…able                                                                                   
Dec 21 20:35:37 haxophone haxo001[558]: [2023-12-21T20:35:36Z INFO  haxo001…357A                                                                                   
Dec 21 20:35:47 haxophone haxo001[558]: Synth created                                                                                                              
Dec 21 20:35:48 haxophone haxo001[558]: [2023-12-21T20:35:48Z INFO  haxo001…idi]                                                                                   
Dec 21 20:35:48 haxophone haxo001[558]:     Available output ports:                                                                                                
Dec 21 20:35:48 haxophone haxo001[558]: [2023-12-21T20:35:48Z INFO  haxo001…14:0                                                                                   
Dec 21 20:35:48 haxophone haxo001[558]: [2023-12-21T20:35:48Z INFO  haxo001…24:0                                                                                   
Dec 21 20:35:48 haxophone haxo001[558]: [2023-12-21T20:35:48Z INFO  haxo001… one                                                                                   
Dec 21 20:35:48 haxophone haxo001[558]: Starting haxophone...                                                                                                      
Hint: Some lines were ellipsized, use -l to show in full.                
```
or run `top`
```
top - 20:36:13 up 1 min,  1 user,  load average: 3.50, 1.40, 0.52
Tasks:  85 total,   2 running,  83 sleeping,   0 stopped,   0 zombie
%Cpu(s): 29.5 us, 47.1 sy,  0.0 ni, 19.4 id,  0.0 wa,  0.0 hi,  4.0 si,  0.0 st
MiB Mem :    429.3 total,     81.1 free,    200.5 used,    147.7 buff/cache
MiB Swap:      0.0 total,      0.0 free,      0.0 used.     68.3 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND    
 561 root      20   0  206376 170660    792 R  68.5  38.8   0:21.33 haxo001   
   11 root      20   0       0      0      0 S   3.2   0.0   0:01.08 ksoftirqd+ 
(...)    
```
4. If we rule out the sensor, that leaves us with the final mystery... why no sound?  You can test just the audio path by stopping the haxo service and playing a file directly with ALSA.  This is how you can do that:
```
pi@haxophone:~$ sudo systemctl stop haxo                                                                                                                           
pi@haxophone:~$ aplay -l                                                                                                                                           
**** List of PLAYBACK Hardware Devices ****                                                                                                                        
card 0: MAX98357A [MAX98357A], device 0: bcm2835-i2s-HiFi HiFi-0 [bcm2835-i2s-HiFi HiFi-0]                                                                         
  Subdevices: 1/1                                                                                                                                                  
  Subdevice #0: subdevice #0                                                                                                                                       
card 1: vc4hdmi [vc4-hdmi], device 0: MAI PCM i2s-hifi-0 [MAI PCM i2s-hifi-0]                                                                                      
  Subdevices: 1/1                                                                                                                                                  
  Subdevice #0: subdevice #0                                                                                                                                                                                                                                                                                                                                                                            
pi@haxophone:~$ aplay /usr/share/sounds/alsa/Side_Right.wav                                                                                                        
Playing WAVE '/usr/share/sounds/alsa/Side_Right.wav' : Signed 16 bit Little Endian, Rate 48000 Hz, Mono 
```
When I do this, I hear the output out of my headphones.  Does this work for you?

**Reply by bberstel** [reply]
I have again good and bad news. 😄 In short: **I have sound**, but not all notes.

What I did:
- I ran `systemctl status` to confirm that the service is running.
- I stopped it and used `aplay` to play sound directly: I couldn't hear the sound.
- I tested my headset again on another sound system to confirm it works. It works, so I was suspecting the line-out port.
- I launched a loop of `aplay` and plugged the headset again. I heard the sound being played while I was plugging the headset and then no more. I unplugged it a very tiny bit and the sound was back. **Conclusion: I was plugging the headset too far away into the line-out port.** (Plus maybe an issue with the I2C port of the first Pi.)
- I restarted the `haxo` service. I could hear the 4-note chime.
- I blew into the haxophone and got no sound, until I realised that my fingers where off by one key so I was pushing L2+L3, which is not a legit fingering. Back on legit ones and I could play a little tune. 👍 

But then:
- I shut the Pi down, unplugged the keyboard and screen, and restarted the Pi.
- I had the 4-note chime.
- And then the only fingerings that (I tried and) gave me some sound where (tenor) middle C# (no key down), high C# (same with octave key), and B. In other words, the only keys that seem to work are L1 and octave, and not together.

I had to stop here. I guess that my next step would be to plug the keyboard and screen again and see whether it re-works, and if so unplug them again and see. I'll try that when I can and I'll keep you posted. Just wanted to tell you the good news, and also the less good one, in case you have ideas.

Best, B

**Reply by jcard0na** [reply]
Oh, wow, this reads like a mystery novel!   Thanks for the frequent updates... (and I'm really sorry that this hasn't been a smooth "plug and play" experience for you... 😮‍💨)

Yes, I have one suggestion... if you had a "stuck open key", you could have some notes that work, while others would not.  Going over [all the fingerings](https://www.sebulli.com/haxo/notemap/) can help identify which stuck open note would produce what you are seeing.  For instance, if L2 key was stuck open, C# and B would work, and all notes below G# would not produce a sound.  

![image](https://github.com/cardonabits/haxo-hw/assets/676181/2988a4a9-13fe-4606-b58c-95a8990eae0f)

But C should work. And A would sound as B.  The high register (D5 and above) would also work.  Can you try comparing A with B, and also check if the high register works?

**Reply by bberstel** [reply]
Victory!

So I started going through all fingerings, well maybe not all of them, but apparently it did un-stuck the key that was stuck, because everything was working properly.

Here is a link to a short video that proves it: https://vimeo.com/898900694 😄 (You'll hear more the keys and my breathing because I simply put the headset next to the phone, but that's ok for the purpose of this video…)

Thanks a lot Javier for your help through this. We got it solved before the end of the year!

Best, B

**Reply by jcard0na** [reply]
Sweet, sweet news!  I'm so happy that we finally made it work.  Thanks a lot for sharing the "proof"!
And just in time for the end of year party celebrations!  (I'm in France to spend "le Réveillon" with friends... you gave me one more reason to celebrate tonight!

Cheers,

**Reply by jcard0na** [reply]
BTW, my father-in-law overheard your recording and said it's Jacques Dutronc, Gentleman Cambrioleur.  Is he right?

**Reply by bberstel** [reply]
Yep, totally right!

Alors je te souhaite un bon réveillon !

Best, B
